bundles/org.openhab.binding.cbus/src/main/java/org/openhab/binding/cbus/internal/discovery/CBusGroupDiscovery.java

```diff
@@ -0,0 +1,104 @@
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.cbus.CBusBindingConstants;
+import org.openhab.binding.cbus.handler.CBusNetworkHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.daveoxley.cbus.Application;
+import com.daveoxley.cbus.CGateException;
+import com.daveoxley.cbus.Group;
+import com.daveoxley.cbus.Network;
+
+/**
+ * The {@link CBusGroupDiscovery} class is used to discover CBus
+ * groups that are in the CBus Network
+ *
+ * @author Scott Linton - Initial contribution
+ */
+@NonNullByDefault
+public class CBusGroupDiscovery extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(CBusGroupDiscovery.class);
+
+    private CBusNetworkHandler cbusNetworkHandler;
+
+    public CBusGroupDiscovery(CBusNetworkHandler cbusNetworkHandler) {
+        super(CBusBindingConstants.SUPPORTED_THING_TYPES_UIDS, 300, false);
+        this.cbusNetworkHandler = cbusNetworkHandler;
+    }
+
+    @Override
+    protected void startScan() {
+        if (cbusNetworkHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
+            try {
+                Map<Integer, ThingTypeUID> applications = new HashMap<Integer, ThingTypeUID>();
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_LIGHTING, CBusBindingConstants.THING_TYPE_LIGHT);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_DALI, CBusBindingConstants.THING_TYPE_DALI);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_TEMPERATURE,
+                        CBusBindingConstants.THING_TYPE_TEMPERATURE);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_TRIGGER,
+                        CBusBindingConstants.THING_TYPE_TRIGGER);
+
+                Network network = cbusNetworkHandler.getNetwork();
+                if (network == null)
+                    return;
+                for (Map.Entry<Integer, ThingTypeUID> applicationItem : applications.entrySet()) {
+                    Application application = network.getApplication(applicationItem.getKey());
+                    if (application == null) {
+                        continue;
+                    }
+                    ArrayList<Group> groups = application.getGroups(false);
+                    for (Group group : groups) {
+                        logger.debug("Found group: {} {} {}", application.getName(), group.getGroupID(),
+                                group.getName());
+                        Map<String, Object> properties = new HashMap<>(4);
```

Comment:
It is probably better to not try to optimize this.
```suggestion
                        Map<String, Object> properties = new HashMap<>();
```

Reply:
Done